ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/GSM/QD.hs

```diff
@@ -0,0 +1,875 @@
+import qualified Ouroboros.Consensus.Node.GSM as GSM
+import           Ouroboros.Consensus.Util.IOLike (IOLike)
+import           Ouroboros.Network.PeerSelection.LedgerPeers.Type
+                     (LedgerStateJudgement (..))
+import           Test.Consensus.GSM.Common
+import qualified Test.QuickCheck as QC
+import           Test.QuickCheck.Gen.Unsafe (Capture (Capture), capture)
+import qualified Test.QuickCheck.Monadic as QC
+import qualified Test.QuickCheck.StateModel as QD
+import           Test.Tasty (TestTree)
+import           Test.Tasty.QuickCheck (testProperty)
+import           Test.Util.Orphans.IOLike ()
+import           Test.Util.Orphans.ToExpr ()
+import           Test.Util.TestEnv (adjustQuickCheckTests)
+import           Test.Util.ToExpr ()
+import qualified Text.PrettyPrint as Pretty
+
+tests :: [TestTree]
+tests = adhoc <> core
+  where
+    adhoc = [ testProperty "GSM yield regression" prop_yield_regression
+            ]
+    core = [
+          adjustQuickCheckTests (* 10)
+        $ testProperty ("GSM (" <> coreTestName upstreamPeerBound <> ")")
+        $ prop_sequential_iosim upstreamPeerBound
+      | upstreamPeerBound <- Nothing : map Just [minBound .. maxBound :: UpstreamPeer]
+      ]
+
+    coreTestName = \case
+        Nothing -> "no peers"
+        Just ub -> "at most " <> case fromEnum ub of
+            0 -> "1 peer"
+            n -> show (n + 1) <> " peers"
+
+-- | Test the example from the Note [Why yield after the command]
+--
+-- This property fails when 'yieldSeveralTimes' is removed/redefined to @pure
+-- ()@.
+prop_yield_regression :: QC.Property
+prop_yield_regression =
+    Reflection.reify (StaticParams{ pUpstreamPeerBound = Nothing
+                                  , pInitialJudgement = YoungEnough
+                                  , pIsHaaSatisfied = const True
+                                  }) $ \(Proxy :: Proxy params) ->
+   QC.once
+ $ runIOSimProp
+ $ prop_sequential_iosim1 @_ @params yieldRegressionActions
+
+ where yieldRegressionActions :: QD.Actions (Model s)
+       yieldRegressionActions =
+            (wrapPositiveAction $ NewCandidate Amara (B 1))
+         <> (wrapPositiveAction $ StartIdling Amara)
+         <> (wrapPositiveAction $ TimePasses 61)
+         <> (wrapPositiveAction $ ExtendSelection (S (-4)))
+         <> (wrapPositiveAction $ ReadMarker)
+
+prop_sequential_iosim :: Maybe UpstreamPeer -> LedgerStateJudgement -> QC.Fun (Set.Set UpstreamPeer) Bool -> QC.Property
+prop_sequential_iosim pUpstreamPeerBound pInitialJudgement (QC.Fun _ pIsHaaSatisfied) =
+    Reflection.reify (StaticParams{ pUpstreamPeerBound
+                                  , pInitialJudgement
+                                  , pIsHaaSatisfied
+                                  }) $ \(Proxy :: Proxy params) ->
+     -- NOTE: the actions have to be generated with an explicit call, as
+     -- 'generator' relies on reflection for access to the parameters
+     QC.forAll QC.arbitrary $ \actions ->
```

Comment:
`QC.forAll` doesn't ever shrink, so our actions would never get shrunk. `QC.property` does the right thing (by using `shrink` from the `Arbitrary` instance).
```suggestion
     QC.property $ \actions ->
```

Reply:
Great catch!